File: solutions/php/strain/1/Strain.php
<?php
declare(strict_types=1);

class Strain
{
    public function keep(array $list, callable $predicate): array
    {
        $keep = [];
        foreach ($list as $item) {
            if ($predicate($item)) {
                $keep[] = $item;
            }
        }
        return $keep;
    }

    public function discard(array $list, callable $predicate): array
    {
        $discard = [];
        foreach ($list as $item) {
            if (!$predicate($item)) {
                $discard[] = $item;
            }
        }
        return $discard;
    }
}

File: solutions/go/strain/1/strain.go
package strain

func Keep[T any](list []T, predicate func(T) bool) []T {
	var result []T
	for _, item := range list {
		if predicate(item) {
			result = append(result, item)
		}
	}
	return result
}

func Discard[T any](list []T, predicate func(T) bool) []T {
	var result []T
	for _, item := range list {
		if !predicate(item) {
			result = append(result, item)
		}
	}
	return result
}
